### ISSUE TITLE
Query: Select-Collection-FirstOrDefault doesn't track entity coming out

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#11762]
```C#
public class Blog
{
    public int Id { get; set; }
    public List<Post> Posts { get; } = new List<Post>();
}

public class Post
{
    public int Id { get; set; }
    public Blog Blog { get; set; }
}

public static void Main(string[] args)
{
    using (var db = new MyContext())
    {
        // Recreate database
        db.Database.EnsureDeleted();
        db.Database.EnsureCreated();

        // Seed database
        db.Add(new Blog
        {
            Posts =
            {
                new Post(),
                new Post(),
                new Post(),
            }
        });

        db.SaveChanges();
    }

    using (var db = new MyContext())
    {
        // Run queries
        var query = db.Blogs.OrderBy(x => x.Id);
        var posts = query.Select(x => x.Posts.OrderBy(m =>
                                           m.Id).FirstOrDefault())
                                  .ToArray();

        //foreach (var message in messages)
        //{ Work-around
            //db.Attach(message);
        //}

        var blogs = query.ToArray();

        Console.WriteLine(db.ChangeTracker.Entries().Count());
    }

    Console.WriteLine("Program finished.");
}
```

Trying to retrieve, blogs and their first posts. We track blogs but we don't track entities coming out of `posts`

**Comment by granthoff1107** [comment]
The work around for this should be. 

query.SelectMany(x => x.Posts.OrderByDescending(p => p.id).Take(1)).Load()

**Comment by smitpatel** [comment]
```C#
        [ConditionalFact]
        public virtual void Take_Select_Navigation()
        {
            AssertQuery<Customer>(
                cs => cs.OrderBy(c => c.CustomerID).Take(2)
                    .Select(c => c.Orders.OrderBy(o => o.OrderID).FirstOrDefault()));
        }

        [ConditionalFact]
        public virtual void Skip_Select_Navigation()
        {
            AssertQuery<Customer>(
                cs => cs.OrderBy(c => c.CustomerID)
                    .Skip(20)
                    .Select(c => c.Orders.OrderBy(o => o.OrderID).FirstOrDefault()),
                assertOrder: true);
        }

        [ConditionalFact]
        public virtual void Select_collection_FirstOrDefault_project_entity()
        {
            AssertQuery<Customer>(
                cs => cs.OrderBy(c => c.CustomerID).Take(2).Select(c => c.Orders.OrderBy(o => o.OrderID).FirstOrDefault()));
        }
```

**Comment by smitpatel** [comment]
@ajcvickers - Making this general issue for tacking with single element result operator.

**Comment by smitpatel** [comment]
This is going to be fixed by using tracking during materialization code path.

**Comment by jwooley** [comment]
It appears this is similar to this scenario. Adding it here for completeness sake and to add another testing scenario. If trying to use a table that doesn't have a primary key and thus someone has accidentally created multiple records with the same "Id". To allow the dictionary creation without duplicate ID's, I did the following:

`var  custs = dc.Customers
                .GroupBy(c => c.SID)
                .Select(c => c.First())
                .ToDictionary(c => c.SID);`

However when trying to check the dc.ChangeTracker.Entities for EntityState.Modified, none of the changes were identified. Fixed by adding a .ToList before the .GroupBy.

(I recognizing the perf penalty that takes and the fact that not having the proper unique constraints on my underlying data, but I'm tracking that change for later.)
